### ISSUE TITLE
container with unless-stopped restart policy is not restarted when the overlay network is not yet available

### ISSUE THREAD

**Issue by sebi-hgdata** [docker/docker#18091]
There seems to be an issue when restarting a container with unless-stopped and an overlay network. (see the inspect error ' "Error": "network v7 not found",' ) Here is my env and steps to reproduce: 
 $ sudo docker info
Containers: 24
Images: 302
Server Version: 1.9.0
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.3.0-040300-generic
Operating System: Ubuntu 14.04.3 LTS
CPUs: 8
Total Memory: 15.16 GiB
Name: sebi
ID: G3MJ:2URQ:P7DQ:5Z4W:GAW4:RYKP:U23J:NPER:IFEQ:XLFB:QX77:H3VA
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Cluster store: consul://localhost:8500
Cluster advertise: 192.168.77.233:2375

$ sudo docker version
Client:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.0
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   76d6bc9
 Built:        Tue Nov  3 17:43:42 UTC 2015
 OS/Arch:      linux/amd64 

$ uname -a
Linux sebi 4.3.0-040300-generic #201511020949 SMP Mon Nov 2 14:50:44 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
1265e5a72ba8        swarm:1.0.0         "/swarm manage --host"   3 days ago          Up 34 minutes                           swarm-manager
3e68afe02fce        swarm:1.0.0         "/swarm join --addr=l"   3 days ago          Up 34 minutes                           swarm-agent
bdfb83da6312        48ea159bfe73        "/bin/sh -c 'sudo dns"   3 days ago          Up 34 minutes                           dnsmasq
60f63d5d58be        195324010cf2        "./start.sh"             3 days ago          Up 34 minutes                           consul-server

$ sudo docker network ls
NETWORK ID          NAME                DRIVER
1e8d37cfbdc6        v7                  overlay             
d352bb53614d        es                  overlay             
0df6681e5e8b        docker_gwbridge     bridge              
9430df44358f        test                bridge              
c7bc60c709f2        bridge              bridge              
70adb2539cf5        none                null                
f46bb02ef887        host                host 

$ sudo docker run -tid --restart="unless-stopped" --net="v7" ubuntu bash 
96108449ddd584567314878b09885014012e21d9107dd21da791a2fa1a4b2f19

$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
96108449ddd5        ubuntu              "bash"                   10 seconds ago      Up 10 seconds                           evil_noyce
1265e5a72ba8        swarm:1.0.0         "/swarm manage --host"   3 days ago          Up 36 minutes                           swarm-manager
3e68afe02fce        swarm:1.0.0         "/swarm join --addr=l"   3 days ago          Up 36 minutes                           swarm-agent
bdfb83da6312        48ea159bfe73        "/bin/sh -c 'sudo dns"   3 days ago          Up 36 minutes                           dnsmasq
60f63d5d58be        195324010cf2        "./start.sh"             3 days ago          Up 36 minutes                           consul-server

$ ps -aef|grep docker\ daemon
adminis+  3623     1  1 17:01 ?        00:00:43 /usr/bin/docker daemon --userland-proxy=false --cluster-advertise=eth0:2375 --cluster-store consul://localhost:8500 -H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock --bip=172.17.42.1/16 --dns 8.8.8.8 --dns 8.8.4.4 -s overlay
seb      22117  3020  0 17:38 pts/26   00:00:00 grep --color=auto docker daemon

$ sudo kill -9 3623

$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
1265e5a72ba8        swarm:1.0.0         "/swarm manage --host"   3 days ago          Up 8 seconds                            swarm-manager
3e68afe02fce        swarm:1.0.0         "/swarm join --addr=l"   3 days ago          Up 8 seconds                            swarm-agent
bdfb83da6312        48ea159bfe73        "/bin/sh -c 'sudo dns"   3 days ago          Up 8 seconds                            dnsmasq
60f63d5d58be        195324010cf2        "./start.sh"             3 days ago          Up 8 seconds                            consul-server

$ sudo docker network ls
NETWORK ID          NAME                DRIVER
1e8d37cfbdc6        v7                  overlay             
d352bb53614d        es                  overlay             
fcb31a4013e6        host                host                
0df6681e5e8b        docker_gwbridge     bridge              
9430df44358f        test                bridge              
eff2051a90e9        bridge              bridge              
d08eb8a89060        none                null 

$ sudo docker inspect evil_noyce
[
{
    "Id": "96108449ddd584567314878b09885014012e21d9107dd21da791a2fa1a4b2f19",
    "Created": "2015-11-19T15:37:40.251001258Z",
    "Path": "bash",
    "Args": [],
    "State": {
        "Status": "exited",
        "Running": false,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 0,
        "ExitCode": 137,
        "Error": "network v7 not found",
        "StartedAt": "2015-11-19T15:37:40.799663568Z",
        "FinishedAt": "2015-11-19T15:38:47.407391317Z"
    },

**Comment by cpuguy83** [comment]
Are you sure this only with `unless-stopped`? Seems like it would be an issue with any attempt to restart before the network is ready.

ping @aboch @mrjana

**Comment by sebi-hgdata** [comment]
Yep, seems like it happens also for `always`(haven't tested for `on-failure` yet)

**Comment by sebi-hgdata** [comment]
Reproduces also for `on-failure`

**Comment by cpuguy83** [comment]
ping @mavenugo

**Comment by mavenugo** [comment]
@sebi-hgdata @cpuguy83 unfortunately, this is expected. if the kv-store is unreachable when the container is started, then the daemon will not be able to connect the container to its network and that results in this issue.
I think the consul-server is also run as a container on the same daemon that is hosting these application containers with the restart policy and hence the consul server is not reachable during daemon restart.

@sebi-hgdata do you see this problem if the consul-server is made reachable during the restart (i.e) running the consul-server container in another host/VM (or) not as a container in the same host ?

**Comment by sebi-hgdata** [comment]
@mavenugo Yes, Consul is run as a container and this does not happen if the overlay network is present, but...
  The issue is with the restart policies, the docker daemon should have kept trying to restart the container, correct?

**Comment by mavenugo** [comment]
@sebi-hgdata thanks for the confirmation. yes, based on the restart policy, the container must be started again (based on the exit-code). am a bit rusty on the restart-policy implementation. I will take a look and confirm the behavior unless @cpuguy83 can help answer that quickly

**Comment by cpuguy83** [comment]
Looks like if there was an error on start it's not attempting restart: https://github.com/docker/docker/blob/master/daemon/monitor.go#L189-L194

**Comment by cpuguy83** [comment]
We can/should add some extra error checks to see if it's a specific issue like, waiting for a linked container to be started or a network to be available.

**Comment by mavenugo** [comment]
@cpuguy83 thats interesting. that will also solve the linked container issue that is haunting many.

**Comment by thaJeztah** [comment]
>  that will also solve the linked container issue that is haunting many.

Also the most discussed issue in Docker Compose https://github.com/docker/compose/issues/374, probably in this case we won't need a check if the *service* is actually usable at that point though.

**Comment by cpuguy83** [comment]
Looking at this, it looks like libnetwork can't handle this situation even if we keep retrying.

This is after restarting the daemon, which has consul running in a container with --restart=always, which the daemon uses for overlay:

```
INFO[0057] POST /v1.22/containers/test3/start
WARN[0057] driver error disconnecting container test3 : could not find network with id b6d7dc8cd7cbc213abd37a92d985bd875bf86249cba75b450c3d3c7f8034f348
WARN[0057] driver error disconnecting container gateway_b500cda515e2 : endpoint not found: b83bed08d1e0eb36b177a02e1f2a9acd1333cc69fdb633fe5bf68463f71af546
WARN[0057] driver error deleting endpoint gateway_b500cda515e2 : endpoint not found: b83bed08d1e0eb36b177a02e1f2a9acd1333cc69fdb633fe5bf68463f71af546
WARN[0057] driver error deleting endpoint test3 : network id "b6d7dc8cd7cbc213abd37a92d985bd875bf86249cba75b450c3d3c7f8034f348" not found
WARN[0057] failed to cleanup ipc mounts:
failed to umount /var/lib/docker/containers/b500cda515e2bd32363e3a2945f6e547ef740e22e7cdf66e524db12f04265838/shm: invalid argument
failed to umount /var/lib/docker/containers/b500cda515e2bd32363e3a2945f6e547ef740e22e7cdf66e524db12f04265838/mqueue: invalid argument
ERRO[0058] Handler for POST /v1.22/containers/test3/start returned error: subnet sandbox join failed for "10.0.0.0/24": error creating vxlan interface: file exists
Error response from daemon: subnet sandbox join failed for "10.0.0.0/24": error creating vxlan interface: file exists
Error: failed to start containers: [test3]
```

These are the errors I see even after the daemon is fully up and running and I try to start the container which is using the overlay.
I can see the overlay, just can't use it, even with a new container:


```
INFO[0174] POST /v1.22/containers/create?name=test3
50b3eda83322ef300fc384324d4e6e0f130027e298da6c3f16b12f568602451f
INFO[0174] POST /v1.22/containers/50b3eda83322ef300fc384324d4e6e0f130027e298da6c3f16b12f568602451f/start
WARN[0174] failed to cleanup ipc mounts:
failed to umount /var/lib/docker/containers/50b3eda83322ef300fc384324d4e6e0f130027e298da6c3f16b12f568602451f/shm: no such file or directory
failed to umount /var/lib/docker/containers/50b3eda83322ef300fc384324d4e6e0f130027e298da6c3f16b12f568602451f/mqueue: no such file or directory
ERRO[0174] Handler for POST /v1.22/containers/50b3eda83322ef300fc384324d4e6e0f130027e298da6c3f16b12f568602451f/start returned error: subnet sandbox join failed for "10.0.0.0/24": error creating vxlan interface: file exists
docker: Error response from daemon: subnet sandbox join failed for "10.0.0.0/24": error creating vxlan interface: file exists.
```

**Comment by mavenugo** [comment]
@cpuguy83 I dont think this is related to container restart condition, but most likely due to the fact that consul lost its content during the restart. Can you please make sure that the consul --data-dir is mounted which guarantees the data being persistent ?
@mrjana @sanimej do you know if any other reasons why we might end up with such an error ?

**Comment by cpuguy83** [comment]
Fixing this for links here: #18208

**Comment by sanimej** [comment]
@mavenugo This error can happen if we try to create a new overlay namespace for the network while the previous one is not cleaned up.. But have to see what sequence can cause this.

@cpuguy83 are you doing an ungraceful restart of the daemon ?

**Comment by cpuguy83** [comment]
@sanimej Nope
It is setup as consul in a container, restart daemon with cluster opts set to point to containerized consul, create overlay network, create container with --net set to the overlay, restart daemon.

**Comment by sebi-hgdata** [comment]
@cpuguy83  I am also, sometimes,  seeing the `subnet sandbox join failed for "10.0.0.0/24": error creating vxlan interface: file exists`  error. after doing some docker service restarts (have to umount and rm /var/run/docker/netns/* to fix the issue )
@mavenugo on my end I have the consul data-dir mounted as a host volume

**Comment by mavenugo** [comment]
I have seen a few issues with running consul as a container especially during restart cases. When a container or daemon restarts, with an existing data dir, consul fails to bootstrap properly when the IP address of the container changes. Can you please check the consul logs to make sure it does come up properly ?

**Comment by sebi-hgdata** [comment]
No issues with Consul after daemon restarts(in my case I used host network for the container). Forgot to add to the steps, but  the container would start successfully if doing a manual start with `sudo docker start evil_noyce `

**Comment by mavenugo** [comment]
@sebi-hgdata just to make sure we are talking about the same problem. The consul issue that I brought up is for the vxlan issue that @cpuguy83 brought up. The restart policy issue is something we have to solve

**Comment by sanimej** [comment]
@cpuguy83 I wasn't able to recreate the problem in the sequence you said.. But I hit the same error accidentally in the following sequence..

- start the container (without cluster options)
- run the consul container
- restart the container with cluster options
- create the overlay network
- create a container on that network. but accidentally network name had a typo. creation failed
- create the container with same name and network; it fails because the previously attempted container still exists
- delete the container and try again. Now I see the error, "error creating vxlan interface: file exists"

I have to add some debugs to find the root cause.. Hopefully the error will be consistent with this sequence..

**Comment by sebi-hgdata** [comment]
@mavenugo just wanted to emphasize that for both issues Consul had no problems

**Comment by sebi-hgdata** [comment]
@cpuguy83 Probably would be a good idea to let the user specify a condition that can be checked before trying to start the container

**Comment by sebi-hgdata** [comment]
This might have been fixed  by #22561, makeing it unreproducible

**Comment by thaJeztah** [comment]
Thanks @sebi-hgdata, let me go ahead and close this, but let us know if you run into this again!
